File: 2023/08/path.rb
class Path
  attr_reader :start, :stop, :current, :directions, :length
  def initialize(start, stop, directions)
    @start, @stop, @directions = start, stop, directions

    @current = start
    @length = 0
  end
  def next_direction
    directions[ length % directions.size ]
  end
  def go_to(node)
    @current = node
    @length += 1
  end
  def done?
    if stop.respond_to? :call
      stop.call current
    else
      stop == current
    end
  end
end

File: 2019/03/Path.java
public class Path {
  private Line current;
  private int length;

  void add(Line line) {
    if(current != null)
      length += current.length();
    current = line;
  }

  public Point intersects(Path path) {
    return current.intersects(path.current);
  }

  public int length(Point to) {
    return this.length + current.length(to);
  }
}
